### PR TITLE
Refactor DB command file

### PR DESCRIPTION
**Pull request by schlessera**
The way options to the underlying `mysql*` tools are currently handled is extremely inconsistent.

This is a first try at making everything more consistent and ensuring that all tools can accept all allowed options.

### CODE REVIEW

**Comment by ChrisWiegman** [comment]
Most of the test failures look to be due to changing the command output by inserting the host, database, user and password params into the command. Updating the test snapshots should be acceptable accordingly. That said, do we need to be passing those params through in such a manner? I'm not sure why these items have changed from what was being done before.

File: src/DB_Command.php [hunk 1]
```diff
@@ -1661,46 +1671,446 @@ private static function get_mysql_args( $assoc_args ) {
...
+			if ( ! array_key_exists( $key, $assoc_args['_db_settings_'] ) ) {
+				WP_CLI::error(
+					"Missing the {$key} information from the database settings.\n" .
+					"It can be set either via the '{$mappings[ 'constant_name' ]}' constant in wp-config.php or via the --{$mappings[ 'flag_name' ]} option."
+				);
+			}
 		}
 
-		return $flag_string;
+		unset( $assoc_args['dbuser'], $assoc_args['dbpass'] );
```
Comment (ChrisWiegman) [inline src/DB_Command.php]: These also should probably be checked that they exist before trying to unset.

File: src/DB_Command.php [hunk 2]
```diff
@@ -1661,46 +1671,446 @@ private static function get_mysql_args( $assoc_args ) {
...
+		if ( array_key_exists( '_db_settings_', $assoc_args ) ) {
+			return $assoc_args['_db_settings_'];
+		}
 
-		if ( array_key_exists( 'defaults', $assoc_args ) ) {
+		$assoc_args['_db_settings_'] = [];
+		if ( isset( $assoc_args['password'] ) && ! isset( $assoc_args['dbpass'] ) ) {
+			$assoc_args['dbpass'] = $assoc_args['password'];
+		}
+		unset( $assoc_args['user'], $assoc_args['pass'], $assoc_args['password'] );
```
[the remark this answers is not in the document]
Comment (schlessera) [reply src/DB_Command.php]: Unsetting is always safe, just like `isset()`.

**Comment by schlessera** [comment]
> I'm not sure why these items have changed from what was being done before.

The goal with the refactor is to normalize the way DB settings are handled and make sure all commands behave in the same way, and this across all environments.

I would think that if passing settings like the user and host explicitly fails, that I'm missing something and I'm not using the correct settings.

**Comment by danielbachhuber** [comment]
Proceeding with https://github.com/wp-cli/wp-cli/issues/5594 for this repository. I've captured this PR to https://gist.github.com/danielbachhuber/02870103195330ca2789ff8aa7b5774a in case this PR is auto-closed or broken in some way.

**Pull request closed without merging**